src/backend/access/transam/xlog.c

```diff
@@ -8968,15 +8966,15 @@ GetLastWrittenLSN(RelFileNode rnode, ForkNumber forknum, BlockNumber blkno)
 /*
  * SetLastWrittenLSNForBlockRange -- Set maximal LSN of written page range.
  * We maintain cache of last written LSNs with limited size and LRU replacement
- * policy. To reduce cache size we store max LSN not for each page, but for
- * bucket (1024 blocks). This cache allows to use old LSN when
- * requesting pages of unchanged or appended relations.
+ * policy. Keeping last written LSN for each page allows to use old LSN when
+ * requesting pages of unchanged or appended relations. Also it is critical for
+ * efficient work of prefetch in case massive update operations (like vacuum or remove).
  *
  * rnode.relNode can be InvalidOid, in this case maxLastWrittenLsn is updated.
  * SetLastWrittenLsn with dummy rnode is used by createdb and dbase_redo functions.
  */
 void
-SetLastWrittenLSNForBlockRange(XLogRecPtr lsn, RelFileNode rnode, ForkNumber forknum, BlockNumber from, BlockNumber till)
+SetLastWrittenLSNForBlockRange(XLogRecPtr lsn, RelFileNode rnode, ForkNumber forknum, BlockNumber from, BlockNumber n_blocks)
```

Comment:
This is a very subtle change!

I was surprised to see that this PR changes only one of the callers of this function. All the other callers look like this:
```
                        SetLastWrittenLSNForBlockRange(XactLastRecEnd,
                                                          index->rd_smgr->smgr_rnode.node,
                                                          MAIN_FORKNUM, 0, RelationGetNumberOfBlocks(index));
```
So before this PR, those callers were actually passing incorrect 'till' value, one too high.

I'd suggest making that bug fix as a separate commit.

Reply:
Well, I didn't consider it is a bug. In previous version of last written lsn cache, it is maintained per chunk. So there is still soe inaccuracy (false positives). So making exclusive high boundary inclusive can cause unnecessary touching of some chunk with probability 1/1024.

Now with per-page last written LSN tracking it becomes more critical. This is why I have made this change. Taken in ccount that adding 1 to REL_METADATA_PSEUDO_BLOCKNO cause overflow I have to change API and (from,length) instead of (from,till).

Reply:
> Well, I didn't consider it is a bug. In previous version of last written lsn cache, it is maintained per chunk. So there is still soe inaccuracy (false positives). So making exclusive high boundary inclusive can cause unnecessary touching of some chunk with probability 1/1024.

Right, it didn't do much harm before, but it was still clearly wrong.

It seems that there's still a lot ot review and discuss with the rest of this PR, so please extract that change, and the shmem allocation fix, into separate PRs. Better to get those out of the way first.

Reply:
Sorry, but it is completely useless for old implementation.
It is doing loop through all affected chunks:
```
		for (bucket = from / LAST_WRITTEN_LSN_CACHE_BUCKET;
			 bucket <= till / LAST_WRITTEN_LSN_CACHE_BUCKET;
			 bucket++)
		{
```
If instead of [from,till) I will pass from,n_blocks
then I still can not use `less-than` comparison instead of `less-or-equal`:
```
		for (bucket = from / LAST_WRITTEN_LSN_CACHE_BUCKET;
			 bucket < (from + n_blocks) / LAST_WRITTEN_LSN_CACHE_BUCKET;
			 bucket++)
		{
```
because of rounding. So with new API this function will do exactly the same as before.

Reply:
> It seems that there's still a lot ot review and discuss with the rest of this PR
Frankly speaking I can not agree with you here.
We can discuss  a lot how current last written LSN tracking is efficient and flexible. And how it can be done in other ways. But I consider this PR mostly as fix or improvement of the current implementation.

The main thing is  reverting my idea of using chunks to minimize size of last written LSN cache. As far as I remember you have proposed it fro the very beginning.  Definitely page-level LSN tracking is more precise and the only drawback is increase cache size. But we in y case are using LRU for replacement.  And for performance most critical are LSNs of recently updated pages which should not be evicted from the cache very fast (LRU).

Concerning `heap_xlog_visible` it is possible to discuss whether it is bug in Postgres or not and how critical it can be. But it is a fact, that such behavior violates page LSN contract and has awful impact on work of our prefetch. Fro the other side - I do not see any negative effect of adding SetPageLSN here.

So from my point of view this PR is doing obvious improvements and fixes. And  delay with its commit will mean that our prefetch and so seqscan will work about two times slower than it can.

Reply:
> Sorry, but it is completely useless for old implementation. It is doing loop through all affected chunks:
> 
> ```
> 		for (bucket = from / LAST_WRITTEN_LSN_CACHE_BUCKET;
> 			 bucket <= till / LAST_WRITTEN_LSN_CACHE_BUCKET;
> 			 bucket++)
> 		{
> ```

If the relation has exactly LAST_WRITTEN_LSN_CACHE_BUCKET blocks, this will visit bucket 0 and 1. It should only visit bucket 0.

> If instead of [from,till) I will pass from,n_blocks then I still can not use `less-than` comparison instead of `less-or-equal`:
> 
> ```
> 		for (bucket = from / LAST_WRITTEN_LSN_CACHE_BUCKET;
> 			 bucket < (from + n_blocks) / LAST_WRITTEN_LSN_CACHE_BUCKET;
> 			 bucket++)
> 		{
> ```
> 
> because of rounding. So with new API this function will do exactly the same as before.

That's not the correct way to the rounding. Here's one way to do it correctly:
```
  BlockNumber start_bucket;
  BlockNumber end_bucket;

  start_bucket = from / LAST_WRITTEN_LSN_CACHE_BUCKET;
  if (n_blocks == 0)
    end_bucket = start_bucket;
  else
  {
    BlockNumber last_block = from + n_blocks - 1;
    end_bucket = (last_block / LAST_WRITTEN_LSN_CACHE_BUCKET) + 1;
  }
  for (bucket = start_bucket; bucket < end_bucket; bucket++)
```

Reply:
>  So from my point of view this PR is doing obvious improvements and fixes. And delay with its commit will mean that our prefetch and so seqscan will work about two times slower than it can.

Yes, it is doing fixes and improvements we obviously need to do. Let's get them in. The way to get them in as quickly as possible is split the PR so that we can focus on the individual changes.